src/compute/src/sink/correction_v2.rs

```diff
@@ -0,0 +1,1109 @@
+    fn split_at_time(self, time: Timestamp) -> (Option<Self>, Option<Self>) {
+        let Some(skip_ts) = time.step_back() else {
+            return (None, Some(self));
+        };
+
+        let before = self.clone();
+        match self.skip_time(skip_ts) {
+            Some((beyond, skipped)) => (before.set_limit(skipped), Some(beyond)),
+            None => (Some(before), None),
+        }
+    }
+
+    /// Attempt to unwrap the cursor into a [`Chain`].
+    ///
+    /// This operation efficiently reuses chunks by directly inserting them into the output chain
+    /// where possible.
+    ///
+    /// An unwrap is only successful if the cursor's `limit` and `overwrite_ts` are both `None` and
+    /// the cursor has unique references to its chunks.
+    fn try_unwrap(self) -> Result<Chain<D>, (&'static str, Self)> {
+        if self.limit.is_some() {
+            return Err(("cursor with limit", self));
+        }
+        if self.overwrite_ts.is_some() {
+            return Err(("cursor with overwrite_ts", self));
+        }
+        if self.chunks.iter().any(|c| Rc::strong_count(c) != 1) {
+            return Err(("cursor on shared chunks", self));
+        }
+
+        let mut chain = Chain::default();
+        let mut remaining = Some(self);
+
+        // We might be partway through the first chunk, in which case we can't reuse it but need to
+        // allocate a new one to contain only the updates the cursor can still yield.
+        while let Some(cursor) = remaining.take() {
+            if cursor.chunk_offset == 0 {
+                remaining = Some(cursor);
+                break;
+            }
+            let update = cursor.get();
+            chain.push(update);
+            remaining = cursor.step();
+        }
+
+        if let Some(cursor) = remaining {
+            for chunk in cursor.chunks {
+                let chunk = Rc::into_inner(chunk).expect("checked above");
+                chain.push_chunk(chunk);
+            }
+        }
+
+        Ok(chain)
+    }
+}
+
+impl<D: Data> From<Cursor<D>> for Chain<D> {
+    fn from(cursor: Cursor<D>) -> Self {
+        match cursor.try_unwrap() {
+            Ok(chain) => chain,
+            Err((_, cursor)) => {
+                let mut chain = Chain::default();
+                chain.push_cursor(cursor);
+                chain
+            }
+        }
+    }
+}
+
+/// A non-empty chunk of updates, backed by a columnation region.
+///
+/// All updates in a chunk are sorted by (time, data) and consolidated.
+///
+/// We would like all chunks to have size [`CHUNK_SIZE_BYTES`], to make it easy for the allocator
+/// to re-use chunk allocations. Unfortunately, the current `TimelyStack`/`ChunkedStack` API
+/// doesn't provide a convenient way to pre-size regions, so chunks are currently only fixed-size
+/// in spirit.
+struct Chunk<D: Data> {
+    /// The contained updates.
+    data: StackWrapper<(D, Timestamp, Diff)>,
```

Comment:
Since `data` has a static capacity, you probably want a different type here. `Array` might be that.

Reply:
Yeah, I considered that briefly. I think the two reasons why I went with the `StackWrapper` were that (a) it already implements `Columnation` and (b) that's what the arrangement merge batcher uses too (well it uses `TimelyStack` because `StackWrapper` is an Mz thing). Would we like the merge batcher to also use `Array` instead?

Reply:
After discussing offline we decided to move to `columnar` instead as a follow-up: https://github.com/MaterializeInc/database-issues/issues/8913